### PR TITLE
feat!: bundle library and emit types

### PR DESCRIPTION
**Pull request by gclapperton**
#### What does this PR do?

This PR bundles the library using Vite into a package that contains the following files
- an ESM file: `./dist/index.js`
- a CJS file: `./dist/index.cjs`
- and the TS type declartions: `./dist/index.d.ts`

All nested packages have been rolled up into these top level files. Most of the contents of `index.ts` have been moved to `policy.ts` and now only the functions exported from `index.ts` are exposed as part of the public API. This better communicates what is public and what is not, as well as allowing us to make internal changes without the risk of breaking any clients that rely on nested packages.

Types are now also emitted for the first time!

#### How should this be manually tested?

Run `npm pack` to bundle the library and then `npm install` the resulting `snyk-policy-0.0.0.tgz` into Registry, CLI or any other consumer of the library to make sure that there are no breaking changes... besides updating/alignment of types.

### CODE REVIEW

File: lib/filter/index.ts [hunk 1]
```diff
@@ -24,8 +24,8 @@ const debug = newDebug('snyk:policy');
  * @param matchStrategy the strategy used to match vulnerabilities (defaults to 'packageManager')
  * @returns the filtered vulnerabilities object
  */
-function filter<T extends Vulnerability>(
-  vulns: VulnerabilityReport<T>,
+function filter<VulnType extends Vulnerability, ReportType>(
+  vulns: ReportType & VulnerabilityReport<VulnType>,
```
Comment (gclapperton) [inline lib/filter/index.ts:28]: The vulnerability object is mutated, so updating the vulns type and the return type to take this into account.

File: lib/index.ts [hunk 2]
```diff
@@ -1,281 +1,4 @@
-import newDebug from 'debug';
```
Comment (gclapperton) [inline lib/index.ts:1]: Most of `index.ts` has been moved to `policy.ts` to better control the exposed API when the library is packaged

File: lib/policy.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
+import newDebug from 'debug';
```
Comment (gclapperton) [inline lib/policy.ts:1]: Moved from `index.ts`

File: vite.config.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        // externalize node dependencies
+        external: ['fs', 'path', 'url', 'util'],
+      },
+  },
+  plugins: [
+    dts({
+      rollupTypes: true,
+    }),
+  ],
+  test: {
```
Comment (gclapperton) [inline vite.config.js:22]: Moved from `vitest.config.ts`

File: package.json [hunk 5]
```diff
@@ -2,16 +2,18 @@
...
   "version": "0.0.0",
   "description": "Snyk's policy parser and matching logic",
-  "files": [
-    "dist/package.json",
-    "dist/lib",
-    "LICENSE",
-    "README.md"
-  ],
-  "main": "./dist/lib/index.js",
+  "type": "module",
```
Comment (pmlyng) [inline package.json:5]: 👍  dual module (esm/cjs), but default to esm.

**Review by pmlyng — APPROVED** [review]
Tried packaging locally and works like a charm ✅

**Review by cat2608 — APPROVED** [review]

**Pull request merged**